### DISCUSSION TITLE
patrol cycle for animated gifs - inverse range not working

### DISCUSSION THREAD

**Discussion by piridium** [ImageMagick/ImageMagick#3096] [Help] [answered]
According to the legacy docs (https://legacy.imagemagick.org/Usage/anim_mods/#patrol), I should be able to create a patrol-cycle animation like so:
```
convert loop.gif -coalesce -duplicate 1,-2-1 -quiet -layers OptimizePlus -loop 0 patrol-cycle.gif
```

Basic duplication works. However, as soon as I reverse the range (second parameter for -duplicate), there's nothing duplicated. I am using the most recent version ImageMagick 7.0.10-56 Q16 x86_64 2021-01-06

Any advice?

Update: It seems that I can't even duplicate frames with indices counted from the end as per the new documentation. `-duplicate 10,-1` should duplicate the last frame ten times. But it's not doing anything.

**Comment by fmw42** [comment]
What are you trying to do with reversing the second argument?  Pleas post your command line after reversing.

**Reply by piridium** [reply]
The command above already has the reversed second argument as per legacy docs.
A forward range would be something like `-duplicate 1,1--2` which duplicates the entire sequence from the second frame with index `1` to the second last frame with index `-2`.
Reversed, it becomes `-duplicate 1,-2-1`which should duplicate from the second last frame `-2` to the second frame `1`.
It seems to have worked in older versions of imagemagick, but apparently does not in the current version.

PS: I need reversing in order to append the frames in reverse order to create a patrol-cycle style effect.

**Comment by GeeMack** [comment]
Trying this with ImageMagick 7.0.10-56 Q16 x64 on Windows 10, this command...

    magick xc:#000 xc:#001 xc:#002 xc:#003 xc:#004 -duplicate 1,1--2 info:

... generates this result as expected, images 0, 1, 2, 3, and 4, followed by images 1, 2, and 3.

    xc:#000[0] XC 1x1 1x1+0+0 16-bit sRGB 0.016u 0:00.003
    xc:#001[1] XC 1x1 1x1+0+0 16-bit sRGB 0.016u 0:00.005
    xc:#002[2] XC 1x1 1x1+0+0 16-bit sRGB 0.016u 0:00.006
    xc:#003[3] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.007
    xc:#004[4] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.008
    xc:#001[5] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.010
    xc:#002[6] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.012
    xc:#003[7] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.014

This command seems to demonstrate the problem...

    magick xc:#000 xc:#001 xc:#002 xc:#003 xc:#004 -duplicate 1,-2-1 info:

Where I might expect it to generate a list of images 0, 1, 2, 3, and 4, followed by images 3, 2, and 1, it produces this result instead, no change...

    xc:#000[0] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.003
    xc:#001[1] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.003
    xc:#002[2] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.005
    xc:#003[3] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.007
    xc:#004[4] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.008

Although it seems to me that we should be able to select a sequence of images, even in reverse, with "-duplicate" or "-clone", it's not really working that way. If that's not the intended behavior of "-clone" and "-duplicate", there is a simple workaround...

This command...

    magick xc:#000 xc:#001 xc:#002 xc:#003 xc:#004 ( -clone 1--2 -reverse ) info:

... generates this result, which appears to be the desired result from the OP.

    xc:#000[0] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.003
    xc:#001[1] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.004
    xc:#002[2] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.006
    xc:#003[3] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.008
    xc:#004[4] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.009
    xc:#003[5] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.010
    xc:#002[6] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.012
    xc:#001[7] XC 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.014

**Reply by piridium** [reply]
This illustrates my findings exactly. And i can confirm your workaround works very well. Thank you!

**The question is, whether that's the intended behavior of "-clone" and "-duplicate".** If you do a search for 'imagemagick gif patrol-cycle' many answers like the legacy behavior come up. [Some of them](https://stackoverflow.com/questions/59718190/animated-gif-reverse-animation-and-append-it-to-the-end-of-the-animation) quite recent. I would argue that it would make a lot of sense that the options continue to behave like they did in ImageMagick v6. Do you think this is a bug? If so, where is the best way to post it?

**Reply by piridium** [reply]
I opened an issue here: https://github.com/ImageMagick/ImageMagick/issues/3113

**Comment by piridium** [comment, marked as answer]
The release 7.0.10-58 fixes this bug. Thanks to @urban-warrior for your great work and thanks to @GeeMack for helping.
